### DISCUSSION TITLE
Issue linking test to static library

### DISCUSSION THREAD

**Discussion by avetiso** [mesonbuild/meson#10054] [Q&A] [answered]
Hey everyone,

New to Meson, running into something that seems like it should be obvious but isn't.

Trying to compile a very simple base64 encoder/decoder:

Top level meson.build:
```
project('base64',
        'cpp',
        version : '1.0.0-5',
        default_options : ['cpp_std=c++17']
        )

subdir('src')
subdir('test')
```

src/ level meson.build:
```
inc = include_directories('base64')

libbase64 = static_library('base64',
                            sources: ['_internal/base64.cpp'],
                            include_directories: inc
                            )
```

test/ level meson.build:
```
exe = executable('test_base64',
                 sources: 'main.cpp',
                 link_with: libbase64
                 )

test('all', exe)
```

Ninja builds the library just fine, but the test executable fails to find the #include <base64/base64.h> header.

I've tried:

Passing the executable the same include_directories() object that I used when compiling the library, install: true when making the library, passing the header as a source to the executable, and a few other things I can't remember because I've been looking at this for too long.

The test file is just a simple main.cpp with a main() function, no gtest or anything else.

It seems like I'm missing something something. What is it?

**Comment by eli-schwartz** [comment, marked as answer]
Since you mention `inc = include_directories('base64')` in the src/ directory, I assume your layout looks like this:

```
.
├── meson.build
├── src
│   ├── base64
│   │   └── base64.h   # <<<
│   ├── _internal
│   │   └── base64.cpp
│   └── meson.build
└── test
    ├── main.cpp
    └── meson.build
```

And base64.cpp is doing `#include "base64.h"` and the static library compiles okay.

But the test program includes the *public interface*, which is `<base64/base64.h>`, which means the include directory you need to add is *different* from the internal one.

I think you need `public_inc = include_directories('.')` as well.


Consider adding a [declare_dependency()](https://mesonbuild.com/Reference-manual_functions.html#declare_dependency) for your library. `libbase64_dep` would link to libbase64, and have '.' as an include directory. Then the test program can just `dependencies: libbase64_dep` and get all of this bundled together.

You may also need a declare_dependency if you intend to use libbase64 as a subproject within another project.

**Reply by avetiso** [reply]
That was it, thank you! I really appreciate you taking the time to help.

**Reply by eli-schwartz** [reply]
No problem, that's what we're here for. :)
